presto-main/src/main/java/com/facebook/presto/metadata/HandleResolver.java

```diff
@@ -169,4 +172,112 @@ public ConnectorHandleResolver resolverFor(String id)
+        private final Optional<Class<? extends ConnectorOutputTableHandle>> outputTableHandle;
+        private final Optional<Class<? extends ConnectorInsertTableHandle>> insertTableHandle;
+        private final Optional<Class<? extends ConnectorPartitioningHandle>> partitioningHandle;
+        private final Optional<Class<? extends ConnectorTransactionHandle>> transactionHandle;
+
+        public MaterializedHandleResolver(ConnectorHandleResolver resolver)
+        {
+            tableHandle = getHandleClass(resolver::getTableHandleClass);
+            layoutHandle = getHandleClass(resolver::getTableLayoutHandleClass);
+            columnHandle = getHandleClass(resolver::getColumnHandleClass);
+            split = getHandleClass(resolver::getSplitClass);
+            indexHandle = getHandleClass(resolver::getIndexHandleClass);
+            outputTableHandle = getHandleClass(resolver::getOutputTableHandleClass);
+            insertTableHandle = getHandleClass(resolver::getInsertTableHandleClass);
+            partitioningHandle = getHandleClass(resolver::getPartitioningHandleClass);
+            transactionHandle = getHandleClass(resolver::getTransactionHandleClass);
+        }
+
+        private static <T> Optional<Class<? extends T>> getHandleClass(Supplier<Class<? extends T>> callable)
+        {
+            try {
+                return Optional.of(callable.get());
+            }
+            catch (UnsupportedOperationException e) {
+                return Optional.empty();
+            }
+        }
+
+        public Optional<Class<? extends ConnectorTableHandle>> getTableHandleClass()
+        {
+            return tableHandle;
+        }
+
+        public Optional<Class<? extends ConnectorTableLayoutHandle>> getTableLayoutHandleClass()
+        {
+            return layoutHandle;
+        }
+
+        public Optional<Class<? extends ColumnHandle>> getColumnHandleClass()
+        {
+            return columnHandle;
+        }
+
+        public Optional<Class<? extends ConnectorSplit>> getSplitClass()
+        {
+            return split;
+        }
+
+        public Optional<Class<? extends ConnectorIndexHandle>> getIndexHandleClass()
+        {
+            return indexHandle;
+        }
+
+        public Optional<Class<? extends ConnectorOutputTableHandle>> getOutputTableHandleClass()
+        {
+            return outputTableHandle;
+        }
+
+        public Optional<Class<? extends ConnectorInsertTableHandle>> getInsertTableHandleClass()
+        {
+            return insertTableHandle;
+        }
+
+        public Optional<Class<? extends ConnectorPartitioningHandle>> getPartitioningHandleClass()
+        {
+            return partitioningHandle;
+        }
+
+        public Optional<Class<? extends ConnectorTransactionHandle>> getTransactionHandleClass()
+        {
+            return transactionHandle;
+        }
+
+        @Override
+        public boolean equals(Object o)
```

Comment:
`equals` is used on line 61, and I figured I should implement `hashCode` as well